src/jit/flowgraph.cpp

```diff
@@ -5768,14 +5768,51 @@ void          Compiler::fgFindBasicBlocks()
 
     if (compIsForInlining())
     {
+        bool hasReturnBlocks = false;
+        bool hasMoreThanOneReturnBlock = false;
+
+        for (BasicBlock* block = fgFirstBB; block != nullptr; block = block->bbNext)
+        {
+            if (block->bbJumpKind == BBJ_RETURN)
+            {
+                if (hasReturnBlocks)
+                {
+                    hasMoreThanOneReturnBlock = true;
+                    break;
+                }
+
+                hasReturnBlocks = true;
+            }
+        }
+
+        if (!compInlineResult->UsesLegacyPolicy())
+        {
+            if (!hasReturnBlocks)
+            {
+                //
+                // Mark the call node as "no return". The inliner might ignore CALLEE_DOES_NOT_RETURN and
+                // fail inline for a different reasons. In that case we still want to make the "no return"
+                // information available to the caller as it can impact caller's code quality.
+                //
+
+                impInlineInfo->iciCall->gtCallMoreFlags |= GTF_CALL_M_DOES_NOT_RETURN;
+            }
+
+            compInlineResult->NoteBool(InlineObservation::CALLEE_DOES_NOT_RETURN, !hasReturnBlocks);
```

Comment:
More like the following I'd say:
```CPP
if (!hasReturnBlocks && !compInlineResult->UsesLegacyPolicy())
{
    // ...
    impInlineInfo->iciCall->gtCallMoreFlags |= GTF_CALL_M_DOES_NOT_RETURN;
}
compInlineResult->NoteBool(InlineObservation::CALLEE_DOES_NOT_RETURN, !hasReturnBlocks);
if (compInlineResult->IsFailure())
{
    return;
}
```
`NoteBool` has to be called no matter what value `hasReturnBlocks` has, otherwise an assert fires in the enhanced policy.